programs/checkout/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+                ctx.accounts.system_program.to_account_info(),
+            ],
+        );
+
+        let bid_orders = &mut ctx.accounts.bid_orders.load_mut()?;
+
+        bid_orders
+            .heap
+            .add(bid_price, ctx.accounts.bidder.to_account_info().key());
+
+        let highest_bid = bid_orders.heap.peek_highest_bid();
+
+        msg!("current highest BID : {}", highest_bid.bid_price);
+
+        Ok(())
+    }
+
+    pub fn cancel_bid(ctx: Context<CancelBid>) -> Result<()> {
+        let bidder = &ctx.accounts.bidder;
+
+        let mut heap = ctx.accounts.bid_orders.load_mut()?;
+
+        // Need a clever way to somehow know the bid price after the let mut heap declaration
+        let bid_price_sol = heap.heap.cancel_bid(bidder.key());
+
+        **ctx
+            .accounts
+            .escrow_sol
+            .to_account_info()
+            .try_borrow_mut_lamports()? -= bid_price_sol;
+        **ctx.accounts.bidder.try_borrow_mut_lamports()? += bid_price_sol;
+
+        // msg!("canceling heap data: {}", heap.heap);
+
+        Ok(())
+    }
+
+    pub fn bid_floor(ctx: Context<BidFloor>) -> Result<()> {
+        msg!("in set_data pubkey");
+
+        let mut heap = ctx.accounts.bid_orders.load_mut()?;
+
+        let highest_bid = heap.heap.pop_highest_bid();
+
+        let mut matched_orders = ctx.accounts.matched_orders.load_mut()?;
+
+        matched_orders.trades.insert_node(highest_bid.bidder_pubkey);
+
+        msg!(
+            "linked holder data {:?}, {:?}",
+            matched_orders.trades.free_head,
+            matched_orders.trades.order_head
+        );
+
+        msg!(
+            "voucher balance user: {}, escorw: {}",
+            ctx.accounts.user_voucher.amount,
+            ctx.accounts.escrow_voucher.amount
+        );
+
+        anchor_spl::token::transfer(
+            CpiContext::new(
+                ctx.accounts.token_program.to_account_info(),
+                anchor_spl::token::Transfer {
+                    from: ctx.accounts.user_voucher.to_account_info(),
+                    to: ctx.accounts.escrow_voucher.to_account_info(),
+                    authority: ctx.accounts.user.to_account_info(),
+                },
+            ),
+            1,
+        )?;
+
+        **ctx
+            .accounts
+            .escrow_sol
+            .to_account_info()
+            .try_borrow_mut_lamports()? -= &highest_bid.bid_price;
+        **ctx.accounts.user.try_borrow_mut_lamports()? += &highest_bid.bid_price;
+
+        Ok(())
+    }
+
+    pub fn fulfill_order(
+        ctx: Context<FulfillOrder>,
+        pubkey_to_remove: Pubkey,
+        checkout_auth_bump: u8,
+    ) -> Result<()> {
+        msg!("in remove_order pubkey: {}", &pubkey_to_remove.to_string());
+
+        let mut matched_orders = ctx.accounts.matched_orders.load_mut()?;
+        matched_orders
+            .trades
+            .remove_node_by_pubkey(pubkey_to_remove);
```

Comment:
Assuming that this always just removes the first node that matches? I feel like better functionality would be to remove multiple. Not a huge issue